RFC-0040-nested-device-mesh.md

```diff
@@ -0,0 +1,123 @@
+
+To address the first limitation, we propose the introduction of a NestedDeviceMesh abstraction.
+This new abstraction allows for more flexibility by allowing nesting of DeviceMesh instances, an element of in a DeviceMesh can be a DeviceMesh, each potentially with different shapes and sizes.
+This would allow users to model more complex parallelism strategies, such as those required by heterogeneous workloads.
+
+To address the second limitation, we propose the process group creation API be rewritten to support dynamic world size by periodically polling the c10d store for the current world size and recreating the nccl communicators in the case of a change.
+
+## **Motivation**
+This proposal is motivated by the need to support dynamic mixed hardware in OpenDiLoCo.
+
+The goal of OpenDiLoCo is to reduce the network requirement of deep learning optimization by utilising an inner and outer optimization.
+Each participating worker will optimize their respective copies of the model with the inner optimization without communicating with other workers.
+The outer optimisation step which requires communication between all workers is only triggered every few hundred steps -- reducing the amount of all worker communications.
+
+OpenDiLoCo plans to support on and off ramping of workers which requires the outer world size to be dynamic.
+It also plans to support the participation of heterogenous workers (e.g. a worker can have 8 H100 or 16 A100) in a single training run which requires the NestedDeviceMesh abstraction.
+
+## **Proposed Implementation**
+The NestedDeviceMesh will extend the existing DeviceMesh class, adding the capability to define a hierarchy of meshes.
+Each level of the hierarchy will represent a different mesh, with the possibility of defining different dimensions and sizes for each mesh.
+
+The implementation of NestedDeviceMesh will also support dynamic world sizes.
+Allowing processes to join and leave the mesh before the collective communication is called.
+
+### Key Components:
+1. **NestedDeviceMesh:**
+   - A NestedDeviceMesh can contain multiple DeviceMesh objects, each with its configuration.
+
+### Code Example
+
+DeviceMesh usage
+```python
+from torch.distributed.device_mesh import init_device_mesh
+mesh_2d = init_device_mesh("cuda", (2, 4), mesh_dim_names=("replicate", "shard"))
+
+# Users can access the underlying process group thru `get_group` API.
+replicate_group = mesh_2d.get_group(mesh_dim="replicate")
+shard_group = mesh_2d.get_group(mesh_dim="shard")
+```
+
+NestedDeviceMesh usage
+```python
+from torch.distributed.device_mesh import init_device_mesh
+
+# Initialize the main outer DeviceMesh
+outer_mesh = init_device_mesh("cuda", (4,), mesh_dim_names=("outer"), rdzv_endpoint="10.0.0.1:29400")
```

Comment:
not sure of mixing rdvz abtraction with the mesh one. IMO the mesh should just read periodically from the store. But the rdvz in itself is a different problem

Reply:
Ah. the idea here is that the submesh could have different rdzv endpoint from the outer mesh. The shape of the mesh however im not so sure if we should include if its allowed to be dynamic.